sympy/core/basic.py

```diff
@@ -1284,9 +1283,25 @@ def has_free(self, *patterns):
         True
         >>> (x + y + 1).has_free(y + 1)
         True
+        >>> (2*x*y + 3).has_free(x*y)
+        True
 
+        Performance is best when passing multiple patterns as a set.
         """
-        return self._has(iterfreeargs, *patterns)
+        if not patterns:
+            return False
+        if len(patterns) == 1:
+            if iterable(patterns[0]):
+                patterns = patterns[0]
+        if not isinstance(patterns, set):
+            patterns = set(patterns)
+        rv = any(a in patterns for a in iterfreeargs(self))
+        if rv:
+            return rv
+        nonsym = [i for i in patterns if not getattr(i, '_diff_wrt')]
```

Comment:
This is why it is slow. If patterns is a large set of symbols then this loop can take longer than the rest of the function. The point about using
```
any(a in patterns for a in iterfreeargs(self))
```
is that if `patterns` is a set of size `N` then the runtime does not depend on `N`: it only depends on the size of `iterfreeargs(self)`.

Reply:
Yes. That is what I meant when I said that I tried to avoid the fallback. But if the first test does not succeed, the second test might *if there are compound symbols*. Do you have a better suggestion of how to handle this? We already know that `return self._has(iterfreeargs, *patterns)` will be slow for a large number of patterns. Won't it be faster if we prune out what we don't have to consider before starting that step?

Reply:
`(x*y + 1).has_free(x*y)` or `(x + y + 1).has_free(x + 1)`

Reply:
I tried a rewrite of `has_free`.

Reply:
I don't think that we need `has_free` to be able to handle cases like that. The common case should not be slowed down in order to handle those cases.

Reply:
I was thinking about that. Maybe a `has_free_subexpr`...but that would need a deprecation since that is the current behavior of `has_free` and is part of the documentation of the function. Can you retest and see if the rewrite is any better?

Reply:
Actually, it seems like `Expr.find` should implement the sub-expression finding that `has_free` currently implements. But `(x + y + 1).find(x + 1)` gives `{}`.

Reply:
The main point is that `has_free_set` suffices for our actual needs. We don't need to break compatibility for `has_free` but we we can have `has_free_set` and keep it simple and focussed.

Reply:
Agreed and done.

Also, regarding `find`: it can be used to easily find generators and subexpressions.
```python
>>> ((x+y+sin(x))/x).find(lambda g:
...  not (g.is_Add or g.is_Mul) and g.has_free_arg({x}))
{x, 1/x, sin(x)}
>>> ((x+y+sin(x))/x).find(lambda g: g.is_Add and g.has_free(x + y))
{x + y + sin(x)}
```